mbuf/mbuf.go

```diff
@@ -4,45 +4,134 @@ package mbuf
 #include <stdint.h>
 #include <rte_config.h>
 #include <rte_mbuf.h>
+char *reset_and_append(struct rte_mbuf *mbuf, void *ptr, size_t len)
+{
+	rte_pktmbuf_reset(mbuf);
+	char *data = rte_pktmbuf_append(mbuf, len);
+	rte_memcpy(data, ptr, len);
+	return data;
+}
+struct rte_mbuf *alloc_reset_and_append(struct rte_mempool *mp, void *ptr, size_t len)
+{
+	struct rte_mbuf *mbuf;
+	mbuf = rte_pktmbuf_alloc(mp);
+	rte_pktmbuf_reset(mbuf);
+	char *data = rte_pktmbuf_append(mbuf, len);
+	if (data == NULL)
+		return NULL;
+	rte_memcpy(data, ptr, len);
+	return mbuf;
+}
+static inline void *get_go_struct(struct rte_mbuf *mbuf, void *ptr, size_t len)
+{
+	char *data_offset;
+	data_offset = rte_pktmbuf_mtod(mbuf, char *);
+	rte_memcpy(ptr, data_offset, len);
+}
 */
 import "C"
 
 import (
-	// "syscall"
+	"errors"
 	"reflect"
+	"syscall"
 	"unsafe"
 
+	"github.com/yerden/go-dpdk/common"
 	"github.com/yerden/go-dpdk/mempool"
 )
 
+var NullData = errors.New("NULL response returned")
+var NotEnoughMbufs = errors.New("not enough entries in the mempool; no mbufs are retrieved")
+
 // Mbuf contains a packet.
 type Mbuf C.struct_rte_mbuf
 
 func mbuf(m *Mbuf) *C.struct_rte_mbuf {
 	return (*C.struct_rte_mbuf)(unsafe.Pointer(m))
 }
 
+func mbufs(ms []*Mbuf) **C.struct_rte_mbuf {
+	sh := (*reflect.SliceHeader)(unsafe.Pointer(&ms))
+	return (**C.struct_rte_mbuf)(unsafe.Pointer(sh.Data))
+}
+
 func mp(m *mempool.Mempool) *C.struct_rte_mempool {
 	return (*C.struct_rte_mempool)(unsafe.Pointer(m))
 }
 
 // PktMbufFree returns this mbuf into its originating mempool along
 // with all its segments.
-func PktMbufFree(m *Mbuf) {
+func (m *Mbuf) PktMbufFree() {
 	C.rte_pktmbuf_free(mbuf(m))
 }
 
 // RawFree returns this mbuf into its originating mempool.
-func RawFree(m *Mbuf) {
+func (m *Mbuf) RawFree() {
 	C.rte_mbuf_raw_free(mbuf(m))
 }
 
 // PktMbufClone clones the mbuf using supplied mempool as the buffer
 // source.
-func PktMbufClone(m *Mbuf, p *mempool.Mempool) {
+func (m *Mbuf) PktMbufClone(p *mempool.Mempool) {
 	C.rte_pktmbuf_clone(mbuf(m), mp(p))
 }
 
+// PktMbufAlloc allocate an uninitialized mbuf from mempool p.
+// Note that NULL may be returned if allocation failed.
+func PktMbufAlloc(p *mempool.Mempool) *Mbuf {
+	m := C.rte_pktmbuf_alloc(mp(p))
+	return (*Mbuf)(m)
+}
+
+// PktMbufAllocBulk allocate a bulk of mbufs.
+func PktMbufAllocBulk(p *mempool.Mempool, ms []*Mbuf) error {
+	e := C.rte_pktmbuf_alloc_bulk(mp(p), mbufs(ms), C.uint(len(ms)))
+	switch syscall.Errno(e) {
+	case 0:
+		return nil
+	case syscall.ENOENT:
+		return NotEnoughMbufs
```

Comment:
return `Errno`; leave interpretation of the error to the caller.

Reply:
fixed